File: Rust/binary-tree-level-order-traversal-ii.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

type Tree = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn level_order_bottom(root: Tree) -> Vec<Vec<i32>> {
        fn search(root: &Tree, level: usize, memo: &mut Vec<Vec<i32>>) {
            root.as_ref().map(|node| {
                if let None = memo.get(level) {
                    memo.push(vec![]);
                }
                let node = node.borrow();
                memo[level].push(node.val);
                search(&node.left , level + 1, memo);
                search(&node.right, level + 1, memo);
            });
        }
        let mut memo: Vec<Vec<i32>> = vec![];
        search(&root, 0, &mut memo);
        memo.into_iter().rev().collect()
    }
}

File: Scala/binary-tree-level-order-traversal-ii.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def levelOrderBottom(root: TreeNode): List[List[Int]] = {
        dfs(root).reverse
    }
    def dfs(root: TreeNode): List[List[Int]] = {
        if (root == null) {
            return Nil
        }
        val lefts  = dfs(root.left)
        val rights = dfs(root.right)
        val merge = {
            var i = 0
            var j = 0
            var xs: List[List[Int]] = List()
            while (i < lefts.length || j < rights.length) {
                val x = if (i >= lefts .length) Nil else lefts(i)
                val y = if (i >= rights.length) Nil else rights(i)
                xs = xs :+ (x ::: y)
                i += 1
                j += 1
            }
            xs
        }
        List(root.value) :: merge
    }
}